### PR TITLE
refactor marcenko-pastur functions into a distribution fit

### PR DESCRIPTION
**Pull request by kkontoudi**

### CODE REVIEW

File: scifin/statistics/distributions.py [hunk 1]
```diff
@@ -1900,48 +1904,146 @@ def cdf(self, k: Union[int, list, range, np.ndarray]) -> Union[float, list, np.n
...
             return cdf
-    
-#---------#---------#---------#---------#---------#---------#---------#---------#---------#
\ No newline at end of file
+
+
+#@typechecked()
+class MarcenkoPastur(Distribution):
+
+    def __init__(self, sigma: float, lambda_: float, name: str = "") -> None:
```
Comment (godfrey-leung) [inline scifin/statistics/distributions.py:1955]: No need for typing hint for return if functions/methods returns nothing

File: tests/test_statistics_distributions.py [hunk 2]
```diff
@@ -431,3 +433,48 @@ def test_methods(self):
...
+        print(sigma1)
+
+        mp_dist = dis.MarcenkoPastur(0.5, 0.1)
+        #X2 = np.diag(X2)
+        print(X2.shape)
+        # Should n_pts be the same as n_features??
+        mp_dist.fit(X2, n_pts=100, bandwidth=0.01, initial_params=[np.sqrt(0.5), 0.1])
+
+        print(mp_dist.sigma)
+        print(mp_dist.lambda_)
```
Comment (godfrey-leung) [inline tests/test_statistics_distributions.py]: why print statements here in the test? Shouldn't they be assert statements testing if outputs as desired?

**Comment by godfrey-leung** [comment]
the `distributions.py` module is too big, with over 2000 lines of codes...would prefer it to be refactored and broken down into separate modules

File: scifin/statistics/distributions.py [hunk 3]
```diff
@@ -1900,48 +1904,146 @@ def cdf(self, k: Union[int, list, range, np.ndarray]) -> Union[float, list, np.n
...
+        """
+        Implements the Probability Density Function (PDF)
+        for the Marcenko-Pastur distribution.
+        """
+
+        lambda_minus = self.sigma ** 2 * (1 - self.lambda_ ** .5) ** 2
+        lambda_plus = self.sigma ** 2 * (1 + self.lambda_ ** .5) ** 2
+
+        # TODO: fix this to output zero if x is not in the [lambda_minus, lambda_plus] interval
+        pdf = ((lambda_plus - x) * (x - lambda_minus)) ** .5 / (2 * self.lambda_ * np.pi * self.sigma ** 2 * x)
```
Comment (FabienNugier) [inline scifin/statistics/distributions.py]: I am a bit confused about this line. Shouldn't we have 'x' instead of 'self.lambda_' in the denominator?

**Review by FabienNugier — COMMENTED** [review]
Very nice new functions!

File: scifin/statistics/distributions.py (hunk 3 above)
Comment (FabienNugier) [reply scifin/statistics/distributions.py]: Okay, now we know it is correct.

**Comment by kkontoudi** [comment]
I fixed the fit convergence. There is still a discrepancy between the old and the new implementation but I know where it comes from.
In the old implementation the sampling space is redefined at each iteration of the optimization whereas in mine the sampling space is kept constant while fitting.

I am in favour of keeping it constant because it doesn't always make sense to vary the range in which we do the fit and also I would like to keep the function generic so that we could use it for all the distributions in the end.

What do you think @FabienNugier ?

**Comment by FabienNugier** [comment]
I agree with your implementation. Congrats also on the writing style, it is very clean!
I agree that extending the fit function to the whole Distribution class is a great idea. Let's do that and be able to fit any of the already implemented distributions (PS: there's still a few of them I should add some day, though it's not a high priority).

**Comment by FabienNugier** [comment]
You can merge.

**Pull request merged**